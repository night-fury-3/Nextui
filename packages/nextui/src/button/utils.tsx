import React from 'react';
import { ButtonProps } from '../button/button';
import { ButtonGroupConfig } from '../button/button-group-context';

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return props;
  return {
    ...props,
    auto: true,
    shadow: false,
    bordered: config.bordered || props.bordered,
    ghost: config.ghost || props.ghost,
    flat: config.flat || props.flat,
    animated: config.animated || props.animated,
    rounded: config.rounded || props.rounded,
    light: config.light || props.light,
    size: config.size || props.size,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  };
};
