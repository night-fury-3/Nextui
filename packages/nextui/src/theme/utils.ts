import { NextUIThemes } from './index';
import useWarning from '@hooks/use-warning';
import { DeepPartial } from '@utils/types';
import { deepMergeObject } from '@utils/object';
import darkTheme from './dark';
import lightTheme from './default';

export type PartialTheme = DeepPartial<NextUIThemes>;

export type ThemeParam =
  | PartialTheme
  | ((theme: PartialTheme) => PartialTheme)
  | undefined;

export const mergeTheme = (
  current: NextUIThemes,
  custom: ThemeParam
): NextUIThemes => {
  if (!custom) return current;
  if (typeof custom === 'function') {
    const merged = custom(current);
    if (!merged || typeof merged !== 'object') {
      useWarning('The theme function must return object value.');
    }
    return merged as NextUIThemes;
  }
  return deepMergeObject<NextUIThemes>(current, custom as NextUIThemes);
};

export const switchTheme = (mergedTheme: NextUIThemes): NextUIThemes => {
  const themes: { [key in NextUIThemes['type']]: NextUIThemes } = {
    light: lightTheme,
    dark: darkTheme,
  };
  return { ...mergedTheme, ...themes[mergedTheme.type] };
};
