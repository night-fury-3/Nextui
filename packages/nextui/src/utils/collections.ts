import React, { ReactNode } from 'react';

export const getId = () => {
  return Math.random().toString(32).slice(2, 10);
};

export const hasChild = (
  children: ReactNode | undefined,
  child: React.ElementType
): boolean => {
  const types = React.Children.map(children, (item) => {
    if (!React.isValidElement(item)) return null;
    return item.type;
  });

  return (types || []).includes(child);
};

export const pick = <Obj extends { [key: string]: any }, Key extends keyof Obj>(
  props: Key[],
  obj: Obj
): Pick<Obj, Key> =>
  props.reduce((acc, prop) => {
    acc[prop] = obj[prop];
    return acc;
  }, {} as Pick<Obj, Key>);

export const pickChild = (
  children: ReactNode | undefined,
  targetChild: React.ElementType
): [ReactNode | undefined, ReactNode | undefined] => {
  let target: ReactNode[] = [];
  const withoutTargetChildren = React.Children.map(children, (item) => {
    if (!React.isValidElement(item)) return item;
    if (item.type === targetChild) {
      target.push(item);
      return null;
    }
    return item;
  });

  const targetChildren = target.length >= 0 ? target : undefined;

  return [withoutTargetChildren, targetChildren];
};
