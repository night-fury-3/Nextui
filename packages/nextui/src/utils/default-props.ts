import { NormalSizes } from './prop-types';
import type { CSSProperties } from 'react';

export type NextUISpacingWithAuto = NormalSizes | 'px' | 'auto' | number;

export interface NextUIMargins {
  m?: NextUISpacingWithAuto;
  my?: NextUISpacingWithAuto;
  mx?: NextUISpacingWithAuto;
  mt?: NextUISpacingWithAuto;
  mb?: NextUISpacingWithAuto;
  ml?: NextUISpacingWithAuto;
  mr?: NextUISpacingWithAuto;
}

export interface NextUIPaddings {
  p?: NextUISpacingWithAuto;
  py?: NextUISpacingWithAuto;
  px?: NextUISpacingWithAuto;
  pt?: NextUISpacingWithAuto;
  pb?: NextUISpacingWithAuto;
  pl?: NextUISpacingWithAuto;
  pr?: NextUISpacingWithAuto;
}

export type NextUISpaces = NextUIMargins & NextUIPaddings;

export type NextUISpacesKeys = { [key in keyof NextUISpaces]?: string };

export const NextUISpacesStringKeys = [
  'm',
  'my',
  'mx',
  'mt',
  'mb',
  'ml',
  'mr',
  'p',
  'py',
  'px',
  'pt',
  'pb',
  'pl',
  'pr'
];

export interface DefaultProps extends NextUIMargins, NextUIPaddings {
  style?: CSSProperties;
  className?: string;
}
