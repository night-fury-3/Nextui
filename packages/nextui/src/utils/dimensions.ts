import { NormalWeights, NormalSizes } from './prop-types';

interface GetSizeValue {
  size: string | number;
  sizes: Record<string, any>;
  defaultSize?: string;
}

export const getMargin = (num: number): string => {
  return `calc(${num * 15.25}pt + 1px * ${num - 1})`;
};

export const getNormalWeight = (weight?: NormalWeights): string | undefined => {
  const weights: { [key in NormalWeights]?: string } = {
    light: '1px',
    normal: '2px',
    bold: '3px'
  };
  return weights[weight || 'normal'];
};

export const getNormalRadius = (
  size: NormalSizes,
  rounded?: boolean
): string => {
  const radius: { [key in NormalSizes]: string } = {
    mini: '7px',
    small: '9px',
    medium: '12px',
    large: '14px',
    xlarge: '15px'
  };
  const baseRadius = radius[size];
  return rounded ? `calc(${baseRadius} + 10rem)` : baseRadius;
};

export function getSizeValue({
  size,
  sizes,
  defaultSize = 'md'
}: GetSizeValue) {
  if (typeof size === 'number') {
    return size;
  }
  return sizes[size] || size || sizes[defaultSize];
}
