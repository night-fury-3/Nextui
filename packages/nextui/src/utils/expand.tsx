import React, { useEffect, useRef, useState } from 'react';
import withDefaults from '../utils/with-defaults';
import useRealShape from '../use-real-shape';
import clsx from './clsx';

interface Props {
  isExpanded?: boolean;
  animated?: boolean;
  delay?: number;
}

const defaultProps = {
  isExpanded: false,
  animated: true,
  delay: 200
};

export type ExpandProps = Props & typeof defaultProps;

const Expand: React.FC<React.PropsWithChildren<ExpandProps>> = ({
  isExpanded,
  delay,
  animated,
  children
}) => {
  const [height, setHeight] = useState<string>(isExpanded ? 'auto' : '0');
  const [selfExpanded, setSelfExpanded] = useState<boolean>(isExpanded);
  const contentRef = useRef<HTMLDivElement>(null);
  const entryTimer = useRef<number>();
  const leaveTimer = useRef<number>();
  const resetTimer = useRef<number>();
  const [state, updateShape] = useRealShape<HTMLDivElement>(contentRef);

  useEffect(() => setHeight(`${state.height}px`), [state.height]);
  useEffect(() => {
    // show element or reset height.
    // force an update once manually, even if the element does not change.
    // (the height of the element might be "auto")
    if (!isExpanded) {
      updateShape();
      setHeight(`${state.height}px`);
    }

    // show expand animation
    entryTimer.current = window.setTimeout(() => {
      setSelfExpanded(isExpanded);
      clearTimeout(entryTimer.current);
    }, 30);

    // Reset height after animation
    if (isExpanded) {
      resetTimer.current = window.setTimeout(() => {
        setHeight('auto');
        clearTimeout(resetTimer.current);
      }, delay);
    } else {
      leaveTimer.current = window.setTimeout(() => {
        clearTimeout(leaveTimer.current);
      }, delay / 2);
    }

    return () => {
      clearTimeout(entryTimer.current);
      clearTimeout(leaveTimer.current);
      clearTimeout(resetTimer.current);
    };
  }, [isExpanded]);

  return (
    <div className={clsx('container', { expanded: selfExpanded, animated })}>
      <div ref={contentRef} className="content">
        {children}
      </div>
      <style jsx>{`
        .container {
          padding: 0;
          margin: 0;
          height: 0;
          opacity: 0;
          overflow: hidden;
        }
        .animated {
          transition: height ${delay}ms ease 0ms,
            opacity ${delay * 1.5}ms ease 0ms;
        }
        .expanded {
          height: ${height};
          opacity: 1;
        }
      `}</style>
    </div>
  );
};

export default withDefaults(Expand, defaultProps);
