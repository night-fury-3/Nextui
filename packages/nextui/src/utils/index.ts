import { NormalColors, normalColors } from './prop-types';
import { NextUIThemesPalette } from '@theme/index';
/**
 * This function allows validate if a string is a hexadecimal
 * value
 * @param str [string] hexadecimal value
 * @returns result [boolean]
 */
export const isHex = (str: string): boolean => {
  const exp = /#[a-fA-F0-9]{3,6}/g;
  return exp.test(str);
};

/**
 *
 * @param hex [string]
 * @param alpha [number]
 * @returns [string]
 */
export const hexToRGBA = (hex: string, alpha: number = 1): string => {
  let r: string | number = 0,
    g: string | number = 0,
    b: string | number = 0;
  // 3 digits
  if (hex.length == 4) {
    r = '0x' + hex[1] + hex[1];
    g = '0x' + hex[2] + hex[2];
    b = '0x' + hex[3] + hex[3];
    // 6 digits
  } else if (hex.length == 7) {
    r = '0x' + hex[1] + hex[2];
    g = '0x' + hex[3] + hex[4];
    b = '0x' + hex[5] + hex[6];
  }
  return `rgba(${+r}, ${+g},${+b},${alpha})`;
};

export const getNormalColor = (
  color: NormalColors | string | undefined,
  palette: NextUIThemesPalette,
  defaultColor: string = 'inherit'
) => {
  const colors: { [key in NormalColors | string]: string } = {
    default: defaultColor,
    primary: palette.primary,
    secondary: palette.secondary,
    success: palette.success,
    warning: palette.warning,
    error: palette.error,
    gradient: palette.gradient,
  };
  return color && colors[color] ? colors[color] : defaultColor;
};

export const isNormalColor = (color: string): boolean => {
  let found = normalColors.find((el) => el === color);
  return found !== undefined && found !== null;
};

/**
 * Function that checks color name support in the current browser
 * @param strColor
 * @returns boolean
 */
export const isColor = (strColor: string) => {
  let s = new Option().style;
  s.color = strColor;
  return s.color == strColor;
};
