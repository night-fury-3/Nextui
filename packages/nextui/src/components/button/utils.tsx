import React, { ReactNode } from 'react';
import { NormalSizes } from '@utils/prop-types';
import ButtonIcon from './button-icon';
import { ButtonProps } from '@components/button/button';
import { ButtonGroupConfig } from '@components/button/button-group-context';

export const getButtonChildrenWithIcon = (
  auto: boolean,
  size: NormalSizes,
  children: ReactNode,
  icons: {
    icon?: React.ReactNode;
    iconRight?: React.ReactNode;
  }
) => {
  const { icon, iconRight } = icons;
  const hasIcon = icon || iconRight;
  const isRight = Boolean(iconRight);
  const paddingForAutoMode =
    auto || size === 'mini'
      ? `calc(var(--next-ui-button-height) / 2 + var(--next-ui-button-padding) * .5)`
      : 0;
  if (!hasIcon) return <div className="text">{children}</div>;
  if (React.Children.count(children) === 0) {
    return (
      <ButtonIcon isRight={isRight} isSingle>
        {hasIcon}
      </ButtonIcon>
    );
  }
  return (
    <React.Fragment>
      <ButtonIcon isRight={isRight}>{hasIcon}</ButtonIcon>
      <div className={`text ${isRight ? 'right' : 'left'}`}>
        {children}
        <style jsx>{`
          .left {
            padding-left: ${paddingForAutoMode};
          }
          .right {
            padding-right: ${paddingForAutoMode};
          }
        `}</style>
      </div>
    </React.Fragment>
  );
};

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return props;
  return {
    ...props,
    auto: true,
    shadow: false,
    bordered: config.bordered || props.bordered,
    flat: config.flat || props.flat,
    animated: config.animated || props.animated,
    rounded: config.rounded || props.rounded,
    light: config.light || props.light,
    size: config.size || props.size,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  };
};
