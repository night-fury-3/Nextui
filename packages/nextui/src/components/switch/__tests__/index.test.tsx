import React from 'react';
import { mount, shallow, ReactWrapper } from 'enzyme';
import { Switch } from '@components';
import { nativeEvent, updateWrapper } from '@tests/utils';

interface IconProps {
  className?: string;
}

const Icon: React.FC<IconProps> = ({ className = '' }) => (
  <svg className={className} />
);

const expectSwitchIsChecked = (wrapper: ReactWrapper) => {
  expect(wrapper.find('.checked').length).not.toBe(0);
};

const expectSwitchIsUnChecked = (wrapper: ReactWrapper) => {
  expect(wrapper.find('.checked').length).toBe(0);
};

describe('Switch', () => {
  it('should render correctly', () => {
    const wrapper = mount(<Switch />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should support square and circle', () => {
    const circle = shallow(<Switch />);
    expect(() => circle.unmount()).not.toThrow();
    const square = shallow(<Switch squared />);
    expect(() => square.unmount()).not.toThrow();
  });

  it('should render correctly with an icon', () => {
    const wrapper = mount(<Switch icon={<Icon />} />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should render correctly with an iconOn', () => {
    const wrapper = mount(<Switch iconOn={<Icon />} />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should render correctly with an iconOff', () => {
    const wrapper = mount(<Switch iconOff={<Icon />} />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should render correctly with both icons on/off', () => {
    const wrapper = mount(<Switch iconOn={<Icon />} iconOff={<Icon />} />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should trigger events when switch changed and change icons on/off', async () => {
    const wrapper = mount(
      <Switch
        initialChecked={false}
        iconOn={<Icon className="icon-on" />}
        iconOff={<Icon className="icon-off" />}
      />
    );

    let icon = wrapper.find('.icon-off').at(0).getDOMNode() as SVGSVGElement;
    expect(icon).not.toBeNull();

    wrapper.find('input').simulate('change', {
      ...nativeEvent,
      target: { checked: true },
    });
    await updateWrapper(wrapper);
    expectSwitchIsChecked(wrapper);

    icon = wrapper.find('.icon-on').at(0).getDOMNode() as SVGSVGElement;
    expect(icon).not.toBeNull();
  });

  it('should throw error when icon is left over', () => {
    let errorMessage = '';
    const errorSpy = jest
      .spyOn(console, 'error')
      .mockImplementation((msg) => (errorMessage = msg));
    mount(<Switch icon={<Icon />} iconOn={<Icon />} />);
    expect(errorMessage).toContain('Remove props "icon"');
    errorSpy.mockRestore();
  });

  it('should work with different colors', () => {
    const wrapper = mount(
      <div>
        <Switch color="primary" />
        <Switch color="secondary" />
        <Switch color="success" />
        <Switch color="warning" />
        <Switch color="error" />
      </div>
    );
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should work with different sizes', () => {
    const wrapper = mount(
      <div>
        <Switch size="mini" />
        <Switch size="small" />
        <Switch size="medium" />
        <Switch size="large" />
        <Switch size="xlarge" />
      </div>
    );
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should set switch follow checked prop', async () => {
    const wrapper = mount(<Switch initialChecked={true} />);
    expectSwitchIsChecked(wrapper);

    wrapper.setProps({ checked: false });
    await updateWrapper(wrapper);
    expectSwitchIsUnChecked(wrapper);

    wrapper.setProps({ checked: true });
    await updateWrapper(wrapper);
    expectSwitchIsChecked(wrapper);
  });

  it('should trigger events when switch changed', async () => {
    let checked = false;
    const changeHandler = jest
      .fn()
      .mockImplementation((e) => (checked = e.target.checked));
    const wrapper = mount(<Switch onChange={changeHandler} />);

    wrapper.find('input').simulate('change', {
      ...nativeEvent,
      target: { checked: true },
    });
    await updateWrapper(wrapper);
    expectSwitchIsChecked(wrapper);

    expect(changeHandler).toHaveBeenCalled();
    expect(checked).toEqual(true);
  });

  it('should ignore events when switch disabled', async () => {
    const changeHandler = jest.fn();
    const wrapper = mount(<Switch onChange={changeHandler} disabled />);

    wrapper.find('input').simulate('change', {
      ...nativeEvent,
      target: { checked: true },
    });
    await updateWrapper(wrapper);
    expectSwitchIsUnChecked(wrapper);

    expect(changeHandler).not.toHaveBeenCalled();
  });
});
