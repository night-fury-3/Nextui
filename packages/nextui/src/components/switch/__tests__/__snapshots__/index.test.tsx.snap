// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render correctly 1`] = `"<label class=\\"jsx-282461663 \\"><input type=\\"checkbox\\" class=\\"jsx-282461663\\"><div class=\\"jsx-282461663 switch  \\"><span class=\\"jsx-282461663 circle\\"></span></div></label>"`;

exports[`Switch should render correctly with an icon 1`] = `"<label class=\\"jsx-282461663 \\"><input type=\\"checkbox\\" class=\\"jsx-282461663\\"><div class=\\"jsx-282461663 switch  \\"><span class=\\"jsx-282461663 circle\\"><svg class=\\"\\"></svg></span></div></label>"`;

exports[`Switch should render correctly with an iconOff 1`] = `"<label class=\\"jsx-282461663 \\"><input type=\\"checkbox\\" class=\\"jsx-282461663\\"><div class=\\"jsx-282461663 switch  \\"><span class=\\"jsx-282461663 circle\\"><svg class=\\"\\"></svg></span></div></label>"`;

exports[`Switch should render correctly with an iconOn 1`] = `"<label class=\\"jsx-282461663 \\"><input type=\\"checkbox\\" class=\\"jsx-282461663\\"><div class=\\"jsx-282461663 switch  \\"><span class=\\"jsx-282461663 circle\\"><svg class=\\"\\"></svg></span></div></label>"`;

exports[`Switch should render correctly with both icons on/off 1`] = `"<label class=\\"jsx-282461663 \\"><input type=\\"checkbox\\" class=\\"jsx-282461663\\"><div class=\\"jsx-282461663 switch  \\"><span class=\\"jsx-282461663 circle\\"><svg class=\\"\\"></svg></span></div></label>"`;

exports[`Switch should work with different colors 1`] = `"<div><label class=\\"jsx-282461663 \\"><input type=\\"checkbox\\" class=\\"jsx-282461663\\"><div class=\\"jsx-282461663 switch  \\"><span class=\\"jsx-282461663 circle\\"></span></div></label><label class=\\"jsx-3547102763 \\"><input type=\\"checkbox\\" class=\\"jsx-3547102763\\"><div class=\\"jsx-3547102763 switch  \\"><span class=\\"jsx-3547102763 circle\\"></span></div></label><label class=\\"jsx-2545493299 \\"><input type=\\"checkbox\\" class=\\"jsx-2545493299\\"><div class=\\"jsx-2545493299 switch  \\"><span class=\\"jsx-2545493299 circle\\"></span></div></label><label class=\\"jsx-1007192776 \\"><input type=\\"checkbox\\" class=\\"jsx-1007192776\\"><div class=\\"jsx-1007192776 switch  \\"><span class=\\"jsx-1007192776 circle\\"></span></div></label><label class=\\"jsx-263406684 \\"><input type=\\"checkbox\\" class=\\"jsx-263406684\\"><div class=\\"jsx-263406684 switch  \\"><span class=\\"jsx-263406684 circle\\"></span></div></label></div>"`;

exports[`Switch should work with different sizes 1`] = `"<div><label class=\\"jsx-212113378 \\"><input type=\\"checkbox\\" class=\\"jsx-212113378\\"><div class=\\"jsx-212113378 switch  \\"><span class=\\"jsx-212113378 circle\\"></span></div></label><label class=\\"jsx-373852355 \\"><input type=\\"checkbox\\" class=\\"jsx-373852355\\"><div class=\\"jsx-373852355 switch  \\"><span class=\\"jsx-373852355 circle\\"></span></div></label><label class=\\"jsx-282461663 \\"><input type=\\"checkbox\\" class=\\"jsx-282461663\\"><div class=\\"jsx-282461663 switch  \\"><span class=\\"jsx-282461663 circle\\"></span></div></label><label class=\\"jsx-2616102616 \\"><input type=\\"checkbox\\" class=\\"jsx-2616102616\\"><div class=\\"jsx-2616102616 switch  \\"><span class=\\"jsx-2616102616 circle\\"></span></div></label><label class=\\"jsx-3908543125 \\"><input type=\\"checkbox\\" class=\\"jsx-3908543125\\"><div class=\\"jsx-3908543125 switch  \\"><span class=\\"jsx-3908543125 circle\\"></span></div></label></div>"`;
